octeon: bring up target itusrouter, itusbridge, itusgateway

Grommish opened the pull request:
Itus Networks Shield - 1Ghz dual-core mips64 / Cavium Octeon 3 SoC, 1Gb RAM, 4Gb eMMC.

Signed-off-by: Donald Hoskins <grommish@gmail.com>

Grommish committed:
octeon: bring up target itusrouter, itusbridge, itusgateway

Itus Networks Shield - 1Ghz dual-core mips64 / Cavium Octeon 3 SoC, 1Gb RAM, 4Gb eMMC,3 GbE 10/100/1000 ports

Information regarding device can be found: https://deviwiki.com/wiki/Itus_Networks_Shield_Pro

Installing OpenWrt on Itus Networks Shield:

1) Identify which mode you wish to install to (Router, Bridge, Gateway) and download the appropriate initramfs
image for the mode
2) On device: mount /dev/mmcblk1p1 /mnt
3) scp openwrt-octeon-itusxxxxx-initramfs-kernel.bin to /mnt/ItusxxxxxImage (ie: /mnt/ItusrouterImage)
4) On device: umount /mnt
5) reboot

Once booted, run the sysupgrade via cli or luCi on the openwrt-octeon-itusxxxxxsrouter-squashfs-sysupgrade.tar file
the mode you are running.

Once rebooted, the system installation is complete.  Your storage partition for the mode is inialized and set.

Signed-off-by: Donald Hoskins <grommish@gmail.com>

Grommish commented:
Built on OpenWrt SNAPSHOT r13871+3-546e140382 / LuCI Master git-20.186.79919-0c47989, kernel 5.4
Tested on device

dwmw2 commented on target/linux/octeon/base-files/lib/preinit/79_move_config:
```diff
@@ -10,6 +10,22 @@ move_config() {
 			[ -f "/mnt/$BACKUP_FILE" ] && mv -f "/mnt/$BACKUP_FILE" /
 			umount /mnt
 			;;
+		itusrouter)
```
Couldn't you have just one case here for `itusrouter|itusbridge|itusgateway` ?

Grommish replied:
Yes, i was splitting them out for testing, but i don't know why I couldn't do a "itus"* there
Will correct

dwmw2 replied:
(from IRC discussion) ... because you actually want them to have different filenames like `/mnt/${BACKUP_FILE}-${board}` to prevent an upgrade intended for one partition being ingested by another partition if the mode switch is flipped at the wrong time.

adrianschmutzler replied:
No itus* please.
Just
```
itusbridge|\
itusgateway|\
itusrouter)
```

Grommish replied:
> No itus* please.
> Just
> 
> ```
> itusbridge|\
> itusgateway|\
> itusrouter)
> ```

No problems.. Will be changed :)

adrianschmutzler commented:
> bring up target itusrouter, itusbridge, itusgateway

"target" -> "device"

That might also be the reason why I partially misunderstand you on IRC.

I'd however just use "octeon: add support for itusrouter, ..." anyway.

adrianschmutzler commented on target/linux/octeon/base-files/etc/board.d/01_network:
```diff
@@ -8,6 +8,9 @@
 board_config_update
 
 case "$(board_name)" in
+itus*)
```
Please don't use wildcards:
https://openwrt.org/docs/guide-developer/device-support-policies#no_wildcards

Grommish replied:
I need to use itus* or define itusrouter itusgateway itusbridge, and they all have the same default port configuration.

Which should I use?

adrianschmutzler replied:
Read the reference. Just replace this by
```
itusbridge|\
itusgateway|\
itusrouter)
```

as already suggested elsewhere (or the more modern names after we have agreed on them).

Grommish replied:
Will correct

adrianschmutzler commented on target/linux/octeon/base-files/lib/preinit/01_sysinfo:
```diff
@@ -1,31 +1,69 @@
-	echo "$name" > /tmp/sysinfo/board_name
-	echo "$machine" > /tmp/sysinfo/model
+        echo "$name" > /tmp/sysinfo/board_name
+        echo "$machine" > /tmp/sysinfo/model
```
Changing these indents is unrelated and shouldn't be done anyway, as we prefer tabs here (traditionally).

Grommish replied:
Will correct

adrianschmutzler commented on target/linux/octeon/base-files/lib/preinit/01_sysinfo:
```diff
@@ -1,31 +1,69 @@
+                 echo "itusrouter"
+           fi
+        fi
+}
```
Is there really no better way?

Actually, if there is no better, I'd actually prefer to make three images, and have the user choose the correct one instead of this detection mechanism.

Grommish replied:
This is in prep to moving these to their own Subtarget, which will not assign BOARD_NAME correctly, or that I couldn't.  This is how the device reads the multi-switch port to know what mode it's in.

dwmw2 replied:
It could be inferred from the `root=` command line that u-boot provides, couldn't it? 

Or we could make u-boot put it somewhere in the device tree?

Grommish replied:
the Initramfs files don't use root=, which is how the images get loaded onto the device.  

This code is what sets /tmp/sysinfo/board_name, so if I can do it here and not have to do with elsewhere in the system..  Or, can someone suggest a better way?

adrianschmutzler replied:
So, please, explain what's actually going on here? Is this one device with three different modes or is it three devices?

Grommish replied:
One device BOARD, but the switch determines which kernel image is chosen at boot, so THREE separate images (separate files, different root= defines, etc)

adrianschmutzler replied:
So, at a reseller where you would buy it, there is actually only one "device" then?

adrianschmutzler replied:
And the three images then just trigger different "configuration modes" for that some physical "device"?

adrianschmutzler commented on target/linux/octeon/base-files/lib/upgrade/platform.sh:
```diff
@@ -28,6 +29,29 @@ platform_copy_config() {
+		mount_root done
+		cp -af "$UPGRADE_BACKUP" "/rom/$BACKUP_FILE"
+		umount /rom
+			;;
```
Please remove your testing leftovers.

Grommish replied:
Will correct

adrianschmutzler commented on target/linux/octeon/base-files/lib/upgrade/platform.sh:
```diff
@@ -28,6 +29,29 @@ platform_copy_config() {
+		cp -af "$UPGRADE_BACKUP" "/mnt/$BACKUP_FILE"
+		umount /mnt
+		;;
+	itusgateway)
```
Same as above, one block with three identifiers, please.

adrianschmutzler commented on target/linux/octeon/base-files/lib/upgrade/platform.sh:
```diff
@@ -38,20 +62,34 @@ platform_do_flash() {
-	mount -t vfat /dev/$kernel /boot
 
-	[ -f /boot/vmlinux.64 -a ! -L /boot/vmlinux.64 ] && {
+	if [[ $board == "itusrouter" || $board == "itusbridge" || $board == "itusgateway" ]]; then
```
No double brackets [[ please, that's bash syntax.

adrianschmutzler commented on target/linux/octeon/base-files/lib/upgrade/platform.sh:
```diff
@@ -68,22 +106,32 @@ platform_do_upgrade() {
 	er | \
-	erlite)
+	erlite | \
+	itus*)
```
No wildcards.

adrianschmutzler commented on target/linux/octeon/image/Makefile:
```diff
@@ -47,4 +47,40 @@ define Device/ubnt_edgerouter-lite
+# Itus Network Shield Profiles
+#
+ITUSROUTER_CMDLINE:=console=ttyS0,115200, root=/dev/mmcblk1p2 rootfstype=squashfs,ext4,f2fs rootwait
+define Device/itusrouter
```
Please use the currently common vendor_model style here, e.g.
itus_shield-router

I'd also be happy if you could use the board name similar to what's done on modern DTS targets, i.e. itus,shield-router based on the suggested name.

adrianschmutzler commented on target/linux/octeon/image/Makefile:
```diff
@@ -47,4 +47,40 @@ define Device/ubnt_edgerouter-lite
+define Device/itusrouter
+  DEVICE_VENDOR := Itus Networks
+  DEVICE_MODEL := Shield Router
+  BOARD_NAME := itusrouter
```
I'm not too familiar with non-DTS targets. Do we still need BOARD_NAME here?

Grommish replied:
it's carried thru the build system as $(board_name) AFAIK, at least, which it isn't a $SUBTARGET, in which case I could only get it to put "itus" (as the subtarget) instead of the BOARD_NAME defined.

adrianschmutzler replied:
subtarget is a completely different thing.

BOARD_NAME is mostly an old mechanism for SUPPORTED_DEVICES and also influences some file names in the image assembly process.

adrianschmutzler commented on target/linux/octeon/patches-5.4/130-itus_shield_support.patch:
```diff
@@ -0,0 +1,42 @@
+ 	case CVMX_BOARD_TYPE_BBGW_REF:
+ 		return "AABCD";
+-	case CVMX_BOARD_TYPE_CUST_DSR1000N:
++	case CVMX_BOARD_TYPE_ITUS_SHIELD:
```
What's the point of renaming that constant?

Grommish replied:
When i did initial bring up, the device reported as CUST_DSR1000N as the board..  So, it was renamed so it would report ITUS_SHIELD correctly

adrianschmutzler replied:
but that's where I don't get it: you are changing the name of the variable here, not the content.

Grommish replied:
If I don't, I know that at least LUCI Shows it CUST_DSR1000N instead of ITUS_SHIELD (CN7020p1.2-1000-AAP) as the Model and Arch

mans0n replied:
Probably the magic is in the `ENUM_BRD_TYPE_CASE`... I'm not sure if this is the best way to deal with it, but I can't think of any better idea.

adrianschmutzler commented on target/linux/octeon/base-files/lib/upgrade/platform.sh:
```diff
@@ -68,22 +106,32 @@ platform_do_upgrade() {
 
 	return 0
-	
+
```
That line can be removed at all.

adrianschmutzler commented:
Please add a more extensive commit description with proper Specifications and Flashing instructions.

mans0n commented:
I'm happy to see a new support PR for Octeon 3. :)
But isn't it supposed to be another subtarget? I thought we should add another line [here](https://github.com/openwrt/openwrt/blob/master/include/target.mk#L174). Does CPU_TYPE in image Makefile really work?

adrianschmutzler commented:
So, to me it appears that the only difference between your three "devices" will be the "root" parameter in cmdline and the partition chosen during upgrade?

This very much sounds to me like this should be one device in OpenWrt after all, and the distinction between the modes would only be used where necessary then.

Grommish commented:
> I'm happy to see a new support PR for Octeon 3. :)
> But isn't it supposed to be another subtarget? I thought we should add another line [here](https://github.com/openwrt/openwrt/blob/master/include/target.mk#L174). Does CPU_TYPE in image Makefile really work?

It did when I had split Itus out as a Subtarget separate from the others, but then I had it in the target.mk.  Should I split out Edgerouter and Itus into their own Subtargets?

Grommish commented:
> So, to me it appears that the only difference between your three "devices" will be the "root" parameter in cmdline and the partition chosen during upgrade?
> 
> This very much sounds to me like this should be one device in OpenWrt after all, and the distinction between the modes would only be used where necessary then.

That would prevent automated builds if the device is included as a device, since any end-user would need to build their own image in order to set the root, no?

adrianschmutzler commented:
> That would prevent automated builds if the device is included as a device, since any end-user would need to build their own image in order to set the root, no?

How is this switched with OEM firmware? Do they provide three different images as well? (still assuming that this is actually one physical device, as this question asked twice has not been answered so far)

mans0n commented:
> Should I split out Edgerouter and Itus into their own Subtargets?

I would suggest `octeonplus` for edgerouter and octeon3 for itus.

> That would prevent automated builds if the device is included as a device, since any end-user would need to build their own image in order to set the root, no?

What is the stock firmware behavior? If you select one mode, then does it reboot to another root partition? Does the bootloader select a root partition depending on the switch mode?

mans0n commented:
Come to think of it, if octeon+ and octeon3 share a same kernel config, then I would not split them into subtargets...

adrianschmutzler commented on target/linux/octeon/base-files/lib/upgrade/platform.sh:
```diff
@@ -38,20 +54,34 @@ platform_do_flash() {
-	mount -t vfat /dev/$kernel /boot
 
-	[ -f /boot/vmlinux.64 -a ! -L /boot/vmlinux.64 ] && {
+	if [ $board == "itusrouter" || $board == "itusbridge" || $board == "itusgateway" ]; then
```
double == are bash, please use single =

adrianschmutzler commented on target/linux/octeon/base-files/lib/preinit/01_sysinfo:
```diff
@@ -1,31 +1,69 @@
-		name="generic"
-		;;
+	      name="generic"
+	      ;;
```
Still indent issues here.

Grommish commented:
It's been decided that the device would need to eliminate 2 of the 3 modes of operation and present only a single image.  Unfortunately, I don't think this will be a viable solution.  I'm going to go ahead and close the PR.
